File: forms/emergency-medical-technician-psychomotor-examination/back-end-with-loco/src/engine/psychomotor_grader.rs
//! Psychomotor grader module.

use super::flagged_issues::detect_additional_flags;
use super::rules::all_rules;
use super::types::{
    AssessmentData, FiredRule, GradingResult, Outcome, PASS_PERCENT_THRESHOLD,
};

/// Grade an EMT psychomotor assessment.
///
/// Logic mirrors the canonical JS engine (`psychomotor-grader.js`):
///   1. Any critical-criterion rule answered "no" → automatic Fail.
///   2. Otherwise, if no items have been answered → Fail (nothing assessed).
///   3. Otherwise, if percent < PASS_PERCENT_THRESHOLD → Fail.
///   4. Otherwise → Pass.
pub fn grade(data: &AssessmentData) -> GradingResult {
    let timestamp = chrono::Utc::now().to_rfc3339();
    let (fired_rules, critical_failures, points, max_points, answered_count, total_rules) =
        run_rules(data);

    let percent = if max_points > 0 {
        (points as f64 / max_points as f64) * 100.0
    } else {
        0.0
    };

    let outcome: Outcome = if !critical_failures.is_empty() {
        "fail".to_string()
    } else if answered_count == 0 {
        "fail".to_string()
    } else if percent < PASS_PERCENT_THRESHOLD {
        "fail".to_string()
    } else {
        "pass".to_string()
    };

    let grading_for_flags = GradingForFlags {
        critical_failures: critical_failures.clone(),
        fired_rules: fired_rules.clone(),
        percent,
        points,
        max_points,
    };
    let additional_flags = detect_additional_flags(data, &grading_for_flags);

    GradingResult {
        outcome,
        points,
        max_points,
        percent,
        critical_failures,
        fired_rules,
        additional_flags,
        answered_count,
        total_rules,
        timestamp,
    }
}

/// Compact bundle passed to the flag detector (subset of GradingResult).
pub struct GradingForFlags {
    /// Critical failures.
    pub critical_failures: Vec<FiredRule>,
    /// Fired rules.
    pub fired_rules: Vec<FiredRule>,
    /// Percent.
    pub percent: f64,
    /// Points.
    pub points: u32,
    /// Max points.
    pub max_points: u32,
}

/// Evaluate every rule. Returns (firedRules, criticalFailures, points,
/// maxPoints, answeredCount, totalRules).
pub fn run_rules(
    data: &AssessmentData,
) -> (Vec<FiredRule>, Vec<FiredRule>, u32, u32, u32, u32) {
    let mut fired_rules = Vec::new();
    let mut critical_failures = Vec::new();
    let mut points: u32 = 0;
    let mut max_points: u32 = 0;
    let mut answered_count: u32 = 0;

    let rules = all_rules();
    let total_rules = rules.len() as u32;

    for rule in &rules {
        let status = (rule.evaluate)(data);
        let mut points_awarded: u32 = 0;

        match status.as_str() {
            "yes" => {
                points_awarded = rule.points;
                points += rule.points;
                max_points += rule.points;
                answered_count += 1;
            }
            "no" => {
                max_points += rule.points;
                answered_count += 1;
            }
            // "na" and "" excluded entirely.
            _ => {}
        }

        let entry = FiredRule {
            id: rule.id.to_string(),
            step: rule.step,
            category: rule.category.to_string(),
            description: rule.label.to_string(),
            critical: rule.critical,
            points: rule.points,
            status: status.clone(),
            points_awarded,
        };

        fired_rules.push(entry.clone());
        if status == "no" && rule.critical {
            critical_failures.push(entry);
        }
    }

    (
        fired_rules,
        critical_failures,
        points,
        max_points,
        answered_count,
        total_rules,
    )
}

File: forms/emergency-medical-technician-psychomotor-examination/front-end-with-svelte/src/lib/engine/psychomotor-grader.ts
// NREMT EMT Psychomotor Skills Examination grader. Pure function: takes an
// `AssessmentData` object, evaluates every declarative rule in the registry,
// composes the prioritised flag list, and returns the pass/fail outcome plus
// the audit trail of fired rules.
//
// NREMT pass/fail logic:
//   - Any critical-criterion rule firing 'no' -> automatic Fail.
//   - Else, points awarded < PASS_PERCENT_THRESHOLD% of maxPoints -> Fail.
//   - Else -> Pass.
//
// Scoring details:
//   - 'yes' awards the rule's `points`; counts toward maxPoints.
//   - 'no'  awards 0;                  counts toward maxPoints.
//   - 'na'  is excluded from both numerator and denominator.
//   - ''    (unanswered) is excluded from both numerator and denominator.

import type { AssessmentData, FiredRule, GradingResult, Outcome } from './types';
import { psychomotorRules, PASS_PERCENT_THRESHOLD } from './rules';
import { detectAdditionalFlags } from './flagged-issues';

/**
 * Evaluate every psychomotor rule, classify the outcome, attach the flag
 * list, and return the audit trail.
 */
export function gradePsychomotor(data: AssessmentData): GradingResult {
	const firedRules: FiredRule[] = [];
	const criticalFailures: FiredRule[] = [];
	let points = 0;
	let maxPoints = 0;
	let answeredCount = 0;

	for (const rule of psychomotorRules) {
		let status: ReturnType<typeof rule.evaluate> = '';
		try {
			status = rule.evaluate(data);
		} catch (e) {
			console.warn(`Psychomotor rule ${rule.id} evaluation failed:`, e);
			status = '';
		}

		let pointsAwarded = 0;
		if (status === 'yes') {
			pointsAwarded = rule.points;
			points += rule.points;
			maxPoints += rule.points;
			answeredCount++;
		} else if (status === 'no') {
			pointsAwarded = 0;
			maxPoints += rule.points;
			answeredCount++;
		}
		// 'na' and '' are excluded entirely.

		const entry: FiredRule = {
			id: rule.id,
			step: rule.step,
			category: rule.category,
			description: rule.label,
			critical: rule.critical,
			points: rule.points,
			status,
			pointsAwarded
		};
		firedRules.push(entry);

		if (status === 'no' && rule.critical) {
			criticalFailures.push(entry);
		}
	}

	const percent = maxPoints > 0 ? (points / maxPoints) * 100 : 0;

	let outcome: Outcome = '';
	if (criticalFailures.length > 0) {
		outcome = 'fail';
	} else if (answeredCount === 0) {
		// Nothing assessed yet — surface a neutral Fail on submission so the
		// examiner is prompted to actually mark items, but don't claim success.
		outcome = 'fail';
	} else if (percent < PASS_PERCENT_THRESHOLD) {
		outcome = 'fail';
	} else {
		outcome = 'pass';
	}

	const result: GradingResult = {
		outcome,
		points,
		maxPoints,
		percent,
		criticalFailures,
		firedRules,
		additionalFlags: [],
		answeredCount,
		totalRules: psychomotorRules.length,
		timestamp: new Date().toISOString()
	};

	result.additionalFlags = detectAdditionalFlags(data, result);

	return result;
}
